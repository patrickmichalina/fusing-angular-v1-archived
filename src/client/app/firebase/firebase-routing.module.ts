import { FirebaseComponent } from './firebase.component'
import { NgModule } from '@angular/core'
import { RouterModule } from '@angular/router'

@NgModule({
  imports: [
    RouterModule.forChild([
      {
        path: '',
        component: FirebaseComponent,
        data: {
          meta: {
            title: 'Firebase'
          }
        }
      }
    ])
  ],
  exports: [RouterModule]
})
export class FirebaseRoutingModule {}
