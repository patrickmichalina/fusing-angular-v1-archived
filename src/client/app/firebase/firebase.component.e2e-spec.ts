import { baseUrl, browser } from '../../../../tools/test/jest.e2e-setup'

describe('Firebase Page', () => {
  it('should have title', async () => {
    const page = browser.goto(`${baseUrl}/firebase`)
    const text = await page.evaluate(() => document.title)
    expect(text).toContain('Firebase')
  })
})
