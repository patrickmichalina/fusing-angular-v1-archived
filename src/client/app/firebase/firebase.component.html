<mat-card>
  <h2>Notes (Firebase Realtime Database</h2>
  <form [formGroup]="noteForm" (ngSubmit)="addNote(noteForm.value.note)">
    <mat-form-field>
      <input matInput placeholder="add your note" formControlName="note">
    </mat-form-field>
    <div>
      <button mat-raised-button color="primary" type="submit">Save</button>
      <button mat-raised-button color="accent" type="button" (click)="clearNotes()">Delete All</button>
    </div>
  </form>
  <ul>
    <li *ngFor="let note of notes$ | async; trackBy: trackByNote">{{ note }}</li>
  </ul>
</mat-card>

<mat-card>
  <h2>Docs (Firebase Firestore)</h2>
  <form [formGroup]="docsForm" (ngSubmit)="addDoc(docsForm.value)">
    <mat-form-field>
      <input matInput placeholder="title" formControlName="title">
    </mat-form-field>
    <mat-form-field>
      <input matInput placeholder="rating" formControlName="rating" type="number">
    </mat-form-field>
    <div>
      <button mat-raised-button color="primary" type="submit">Save</button>
      <button mat-raised-button color="accent" type="button" (click)="clearDocs()">Delete All</button>
    </div>
  </form>
  <ul>
    <li *ngFor="let doc of docs$ | async; trackBy: trackByDoc">
      <p>Title: {{ doc.title }}</p>
      <p>Rating: {{ doc.rating }}</p>
    </li>
  </ul>
</mat-card>