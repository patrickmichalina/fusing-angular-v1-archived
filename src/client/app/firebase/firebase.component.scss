:host {
  display: block;
  mat-card {
    margin-bottom: 1em;
  }
  form {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }
}
