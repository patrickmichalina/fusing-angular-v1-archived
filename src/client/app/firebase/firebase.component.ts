import { ChangeDetectionStrategy, Component } from '@angular/core'
import { UniversalRtDbService } from '../shared/firebase/firebase-rtdb.service'
import { UniversalFirestoreService } from '../shared/firebase/firebase-firestore.service'
import { FormControl, FormGroup, Validators } from '@angular/forms'

@Component({
  selector: 'pm-firebase',
  templateUrl: './firebase.component.html',
  styleUrls: ['./firebase.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class FirebaseComponent {
  readonly notes$ = this.rtdb.serverCachedListValueChanges('notes', ref =>
    ref.limitToLast(3)
  )
  readonly rtdbObj$ = this.rtdb.serverCachedObjectValueChanges(
    'site-settings/i18n'
  )
  readonly docs$ = this.fs.serverCachedCollectionValueChanges('docs')

  readonly noteForm = new FormGroup({
    note: new FormControl(undefined, [Validators.required])
  })

  readonly docsForm = new FormGroup({
    title: new FormControl(undefined, [Validators.required]),
    rating: new FormControl(1, [])
  })

  constructor(
    private rtdb: UniversalRtDbService,
    private fs: UniversalFirestoreService
  ) {}

  addNote(note: string) {
    this.rtdb.angularFireDatabase.list('notes').push(note)
  }

  clearNotes() {
    this.rtdb.angularFireDatabase.list('notes').remove()
  }

  addDoc(doc: any) {
    this.fs.afs.collection('docs').add(doc)
  }

  clearDocs() {
    // TODO: this.fs.afs.collection('docs').delete()
  }

  trackByDoc(index: number, item: any) {
    return item.title
  }

  trackByNote(index: number, item: string) {
    return item
  }
}
