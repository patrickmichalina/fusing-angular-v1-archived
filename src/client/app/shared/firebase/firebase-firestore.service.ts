import { catchError, distinctUntilChanged, startWith } from 'rxjs/operators'
import { Inject, Injectable, InjectionToken } from '@angular/core'
import { makeStateKey, TransferState } from '@angular/platform-browser'
import { sha1 } from 'object-hash'
import { of } from 'rxjs'
import { AngularFirestore, QueryFn } from 'angularfire2/firestore'

export const FIREBASE_FIRESTORE_TS_PREFIX = new InjectionToken<string>(
  'cfg.fb.fs.prefix'
)

@Injectable()
export class UniversalFirestoreService {
  constructor(
    private ts: TransferState,
    public afs: AngularFirestore,
    @Inject(FIREBASE_FIRESTORE_TS_PREFIX) private prefix: string
  ) {}

  serverCachedDocValueChanges<T>(path: string) {
    const cached = this.ts.get<T | undefined>(this.cacheKey(path), undefined)
    return this.afs
      .doc<T>(path)
      .valueChanges()
      .pipe(
        startWith(cached as T),
        distinctUntilChanged((x, y) => sha1(x) === sha1(y)),
        catchError(err => of(cached))
      )
  }

  serverCachedCollectionValueChanges<T>(path: string, queryFn?: QueryFn) {
    const cached = this.ts.get<ReadonlyArray<T> | undefined>(
      this.cacheKey(path),
      []
    )
    return this.afs
      .collection<T>(path, queryFn)
      .valueChanges()
      .pipe(
        startWith(cached),
        distinctUntilChanged((x, y) => sha1(x) === sha1(y)),
        catchError(err => of(cached))
      )
  }

  cacheKey(path: string) {
    return makeStateKey<string>(`${this.prefix}.${path}`)
  }
}
