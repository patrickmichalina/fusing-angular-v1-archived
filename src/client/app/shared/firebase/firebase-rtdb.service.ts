import {
  catchError,
  distinctUntilChanged,
  startWith,
  tap
} from 'rxjs/operators'
import {
  ApplicationRef,
  Inject,
  Injectable,
  InjectionToken
} from '@angular/core'
import { AngularFireDatabase } from 'angularfire2/database'
import { makeStateKey, TransferState } from '@angular/platform-browser'
import { sha1 } from 'object-hash'
// tslint:disable-next-line:import-blacklist
import { Observable, of } from 'rxjs'
import { Query, Reference } from '@firebase/database-types'

export const FIREBASE_RTDB_TS_PREFIX = new InjectionToken<string>(
  'cfg.fb.rtdb.prefix'
)

const piper = (failureReturn?: any) => <T>(source: Observable<T>) =>
  source.pipe(
    distinctUntilChanged((x, y) => sha1(x) === sha1(y)),
    catchError(err => of(failureReturn))
  )
const cachedPiper = <T>(cached: T) => (source: Observable<T>) =>
  source.pipe(startWith(cached), piper())

@Injectable()
export class UniversalRtDbService {
  // tslint:disable-next-line:readonly-keyword
  readFromCache = true
  constructor(
    public angularFireDatabase: AngularFireDatabase,
    private ts: TransferState,
    appRef: ApplicationRef,
    @Inject(FIREBASE_RTDB_TS_PREFIX) private prefix: string
  ) {
    // tslint:disable-next-line:no-object-mutation
    // appRef.isStable.pipe(filter(Boolean), take(1)).subscribe(() => this.readFromCache = false)
  }

  private turnOffCache() {
    // tslint:disable-next-line:no-object-mutation
    this.readFromCache = false
  }

  serverCachedObjectValueChanges<T>(path: string) {
    const base = this.angularFireDatabase.object<T>(path).valueChanges()
    return this.valueChangesGen<T | null>(path, base)
  }

  serverCachedListValueChanges<T>(
    path: string,
    queryFn?: (ref: Reference) => Query
  ) {
    const base = this.angularFireDatabase.list<T>(path, queryFn).valueChanges()
    return this.valueChangesGen<ReadonlyArray<T>>(path, base, [])
  }

  private valueChangesGen<T>(
    path: string,
    obs: Observable<T>,
    failureReturn?: any
  ) {
    const cached =
      (this.readFromCache &&
        this.ts.get<T | undefined>(this.cacheKey(path), undefined)) ||
      undefined

    return cached
      ? obs.pipe(cachedPiper(cached), tap(() => this.turnOffCache()))
      : obs.pipe(piper(failureReturn))
  }

  private cacheKey(path: string) {
    return makeStateKey<string>(`${this.prefix}.${path}`)
  }
}
