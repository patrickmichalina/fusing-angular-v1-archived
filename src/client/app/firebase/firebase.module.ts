import { FirebaseRoutingModule } from './firebase-routing.module'
import { FirebaseComponent } from './firebase.component'
import { NgModule } from '@angular/core'
import { SharedModule } from '../shared/shared.module'

@NgModule({
  imports: [FirebaseRoutingModule, SharedModule],
  declarations: [FirebaseComponent],
  exports: [FirebaseComponent]
})
export class FirebaseModule {}
