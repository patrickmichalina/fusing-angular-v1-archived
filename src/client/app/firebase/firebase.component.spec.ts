import { FirebaseComponent } from './firebase.component'
import { TransferState } from '@angular/platform-browser'
import { async, ComponentFixture, TestBed } from '@angular/core/testing'
import { Component } from '@angular/core'
import { FirebaseModule } from './firebase.module'
import { AppTestingModule } from '../../../testing/app-testing.module'

@Component({
  selector: 'test-component',
  template: '<pm-firebase></pm-firebase>'
})
class TestComponent {}

describe(FirebaseComponent.name, () => {
  let fixture: ComponentFixture<TestComponent>

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        imports: [FirebaseModule, AppTestingModule.forRoot()],
        declarations: [TestComponent],
        providers: [TransferState]
      }).compileComponents()
    })
  )

  beforeEach(
    async(() => {
      fixture = TestBed.createComponent(TestComponent)
    })
  )

  afterEach(
    async(() => {
      TestBed.resetTestingModule()
    })
  )

  it(
    'should compile',
    async(() => {
      fixture.detectChanges()
      expect(fixture.componentInstance).toBeDefined()
      expect(fixture).toMatchSnapshot()
    })
  )
})
