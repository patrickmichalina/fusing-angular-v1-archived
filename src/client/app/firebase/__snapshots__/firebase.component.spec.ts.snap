// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseComponent should compile 1`] = `
<test-component>
  <pm-firebase>
    <mat-card
      class="mat-card"
    >
      <h2>
        Notes (Firebase Realtime Database
      </h2>
      <form
        class="ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <mat-form-field
          class="mat-input-container mat-form-field ng-tns-c1-0 mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-invalid"
        >
          <div
            class="mat-input-wrapper mat-form-field-wrapper"
          >
            <div
              class="mat-input-flex mat-form-field-flex"
            >
              
              <div
                class="mat-input-infix mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid"
                  formcontrolname="note"
                  id="mat-input-0"
                  matinput=""
                  ng-reflect-name="note"
                  ng-reflect-placeholder="add your note"
                  placeholder="add your note"
                />
                <span
                  class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-0"
                    class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c1-0 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-0"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    add your note
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            <div
              class="mat-input-underline mat-form-field-underline"
            >
              <span
                class="mat-input-ripple mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity: 1;"
              >
                
                <div
                  class="mat-input-hint-spacer mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <div>
          <button
            class="mat-raised-button mat-primary"
            color="primary"
            mat-raised-button=""
            ng-reflect-color="primary"
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
              Save
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
          <button
            class="mat-raised-button mat-accent"
            color="accent"
            mat-raised-button=""
            ng-reflect-color="accent"
            type="button"
          >
            <span
              class="mat-button-wrapper"
            >
              Delete All
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
      </form>
      <ul>
        
      </ul>
    </mat-card>
    <mat-card
      class="mat-card"
    >
      <h2>
        Docs (Firebase Firestore)
      </h2>
      <form
        class="ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <mat-form-field
          class="mat-input-container mat-form-field ng-tns-c1-1 mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-invalid"
        >
          <div
            class="mat-input-wrapper mat-form-field-wrapper"
          >
            <div
              class="mat-input-flex mat-form-field-flex"
            >
              
              <div
                class="mat-input-infix mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid"
                  formcontrolname="title"
                  id="mat-input-1"
                  matinput=""
                  ng-reflect-name="title"
                  ng-reflect-placeholder="title"
                  placeholder="title"
                />
                <span
                  class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-1"
                    class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c1-1 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-1"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    title
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            <div
              class="mat-input-underline mat-form-field-underline"
            >
              <span
                class="mat-input-ripple mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity: 1;"
              >
                
                <div
                  class="mat-input-hint-spacer mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <mat-form-field
          class="mat-input-container mat-form-field ng-tns-c1-2 mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-should-float mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-valid"
        >
          <div
            class="mat-input-wrapper mat-form-field-wrapper"
          >
            <div
              class="mat-input-flex mat-form-field-flex"
            >
              
              <div
                class="mat-input-infix mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid"
                  formcontrolname="rating"
                  id="mat-input-2"
                  matinput=""
                  ng-reflect-name="rating"
                  ng-reflect-placeholder="rating"
                  ng-reflect-type="number"
                  placeholder="rating"
                  type="number"
                />
                <span
                  class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-2"
                    class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c1-2 ng-star-inserted"
                    for="mat-input-2"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    rating
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            <div
              class="mat-input-underline mat-form-field-underline"
            >
              <span
                class="mat-input-ripple mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-2 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity: 1;"
              >
                
                <div
                  class="mat-input-hint-spacer mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <div>
          <button
            class="mat-raised-button mat-primary"
            color="primary"
            mat-raised-button=""
            ng-reflect-color="primary"
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
              Save
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
          <button
            class="mat-raised-button mat-accent"
            color="accent"
            mat-raised-button=""
            ng-reflect-color="accent"
            type="button"
          >
            <span
              class="mat-button-wrapper"
            >
              Delete All
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
      </form>
      <ul>
        
      </ul>
    </mat-card>
  </pm-firebase>
</test-component>
`;
